### PR TITLE
Update develop-csharp-plugins.md

### PR DESCRIPTION
**Pull request by taooceros**
Add document for c# plugin development

TODO:
- [x] API document for Flow.Launcher.Plugin
- [x] Document for plugin.json

### CODE REVIEW

**Comment by taooceros** [comment]
This is my first time writing a document, and I am not good at it😂.

Any suggestion? @Flow-Launcher/team

**Comment by pc223** [comment]
Nice! These will helps people make more plugins for Flow. I suggest do a kind of walkthrough tutorial, with pictures, like "How to build a simple C# plugin for Flow".

I'm thinking of writing a tutorial like that for Python, but still haven't got the time 😄

**Review by deefrawley — DISMISSED** [review]
Nice! I've corrected a few spelling errors and some wording that didn't seem quite right.

**Comment by taooceros** [comment]
@Zeroto521 Could you take a check whether I have rearranged the sidebar correctly?

File: Flow.Launcher.Plugin/_sidebar.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
+# Flow.Launcher.Plugin
```
Comment (Zeroto521) [inline Flow.Launcher.Plugin/_sidebar.md:1]: remove title

File: Flow.Launcher.Plugin/_sidebar.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    - [Result](./flow.launcher.plugin.result.md)
+    - [ResultUpdatedEventArgs](./flow.launcher.plugin.resultupdatedeventargs.md)
+    - [SpecialKeyState](./flow.launcher.plugin.specialkeystate.md)
+- Flow.Launcher.Plugin.SharedCommands
+    - [FilesFolders](./flow.launcher.plugin.sharedcommands.filesfolders.md)
+    - [SearchWeb](./flow.launcher.plugin.sharedcommands.searchweb.md)
+    - [ShellCommand](./flow.launcher.plugin.sharedcommands.shellcommand.md)
+- Flow.Launcher.Plugin.SharedModels
+    - [MatchResult](./flow.launcher.plugin.sharedmodels.matchresult.md)
+    - [SearchPrecisionScore](./flow.launcher.plugin.sharedmodels.searchprecisionscore.md)
```
Comment (Zeroto521) [inline Flow.Launcher.Plugin/_sidebar.md:32]: How to handle this.

## build three folders

- `Flow.Launcher.Plugin`
- `Flow.Launcher.Plugin.SharedCommands`
- `Flow.Launcher.Plugin.SharedModels`

Each folder has its own `_sidebar.md`

## One folder, and hold three toc (table of content) files.

- `Flow.Launcher.Plugin.md`
- `Flow.Launcher.Plugin.SharedCommands.md`
- `Flow.Launcher.Plugin.SharedModels.md`

Let `_sidebar.md` links to these three toc file. And these toc files link to MD file.

File: Flow.Launcher.Plugin/flow.launcher.plugin.actioncontext.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
+# ActionContext
```
Comment (Zeroto521) [inline Flow.Launcher.Plugin/flow.launcher.plugin.actioncontext.md:1]: The title level should start from level two.

File: Flow.Launcher.Plugin/index.md [hunk 4]
```diff
@@ -0,0 +1,61 @@
+# Flow.Launcher.Plugin
+
+[ActionContext](./flow.launcher.plugin.actioncontext.md)
```
Comment (Zeroto521) [inline Flow.Launcher.Plugin/index.md:3]: `./xx.md` has some problem, it got 404.

File: _sidebar.md [hunk 5]
```diff
@@ -1,7 +1,10 @@
...
 - [**Usage Tips**](/usage-tips.md)
-- [**JSON RPC**](/json-rpc.md)
 - [**Plugin Directory**](/plugins.md)
-- [**Porting Plugins**](/port-plugins.md)
-- [**Develop C# Plugins**](/develop-csharp-plugins.md)
-- [**Develop Python Plugins**](/develop-py-plugins.md)
+- Plugin Development
+    - [**plugin.json**](/plugins.json.md)
+    - [**JSON RPC**](/json-rpc.md)
+        - [**Develop Python Plugins**](/develop-py-plugins.md)
```
Comment (Zeroto521) [inline _sidebar.md:7]: click https://zeroto521.github.io/docs/#/json-rpc and https://zeroto521.github.io/docs/#/develop-py-plugins.

Compare to the sidebar changes.

You would see what I have said at up.

**Review by Zeroto521 — DISMISSED** [review]
Well, The HTML rendering result is a total mess, especial in the sidebar.
See https://zeroto521.github.io/docs/#/

Something about the sidebar.
Docsify would find the title automatically except the 1st level title (or called the entry).
We just give the link to the file.
The files are no level different at all.

File: plugin.json.md [hunk 6]
```diff
@@ -0,0 +1,19 @@
+## plugin.json
```
Comment (Zeroto521) [inline plugin.json.md:1]: I thought this file could simply link to https://github.com/Flow-Launcher/Flow.Launcher.PluginsManifest.

File: develop-csharp-plugins.md [hunk 7]
```diff
@@ -1,7 +1,66 @@
-## How to develop C# plugins
+# C#/F# plugins Development Guide
```
Comment (Zeroto521) [inline develop-csharp-plugins.md:1]: `/` can't render well.

**Comment by taooceros** [comment]
Thank you for reviewing! I will modify it now.

File: _sidebar.md (hunk 5 above)
Comment (taooceros) [reply _sidebar.md:7]: Got you

File: plugin.json.md (hunk 6 above)
Comment (taooceros) [reply plugin.json.md:1]: The manifest file is slightly different from the plugin.json, and the document kindly suggests that people already have the plugin.json.

**Comment by taooceros** [comment]
By the way, I think vue theme looks cleaner than bubble theme. You can check the effect here.
https://taooceros.github.io/docs/#/

File: develop-csharp-plugins.md [hunk 8]
```diff
@@ -1,7 +1,71 @@
...
+Flow is written in C#, so plugins written in dotnet platform can directly communicate with Flow without extra protocols.
 
-For now you can use the [dotnet template](https://github.com/Flow-Launcher/dotnet-template) to generate your plugin 
+## Initialization
+
+We recommand you use the [dotnet template](https://github.com/Flow-Launcher/dotnet-template) to generate a plugin template first.
+
+In order to be recongized as a Flow DotNet plugin, the directory need to have at least two files
+1. [`plugin.json`](plugin.json.md)
+2. A Dotnet class that implement **[IPlugin](API-Reference/Flow.Launcher.Plugin/iplugin.md)** or **[IAsyncPlugin](API-Reference/Flow.Launcher.Plugin/iasyncplugin.md)** (remember to refrence [Flow.Launcher.Plugin](https://www.nuget.org/packages/Flow.Launcher.Plugin/) by Nuget). The plugin template will add the reference and create  a`Main.cs` that implement `IPlugin`.
```
Comment (JohnTheGr8) [inline develop-csharp-plugins.md:11]: change "implement" to "implements" (twice)

also fix the spacing for "create  a`Main.cs`"

File: develop-csharp-plugins.md [hunk 9]
```diff
@@ -1,7 +1,71 @@
...
+
+Found our API Reference [here](API-Reference/)
```
Comment (JohnTheGr8) [inline develop-csharp-plugins.md:13]: "Found" -> "Find"

File: develop-csharp-plugins.md [hunk 10]
```diff
@@ -1,7 +1,71 @@
...
 
-We are currently writing up this section, will be updated soon.
+Flow is written in C#, so plugins written in dotnet platform can directly communicate with Flow without extra protocols.
 
-For now you can use the [dotnet template](https://github.com/Flow-Launcher/dotnet-template) to generate your plugin 
+## Initialization
+
+We recommand you use the [dotnet template](https://github.com/Flow-Launcher/dotnet-template) to generate a plugin template first.
+
+In order to be recongized as a Flow DotNet plugin, the directory need to have at least two files
```
Comment (JohnTheGr8) [inline develop-csharp-plugins.md:9]: "need" -> "needs"

File: develop-csharp-plugins.md [hunk 11]
```diff
@@ -1,7 +1,71 @@
...
+1. [`plugin.json`](plugin.json.md)
+2. A Dotnet class that implement **[IPlugin](API-Reference/Flow.Launcher.Plugin/iplugin.md)** or **[IAsyncPlugin](API-Reference/Flow.Launcher.Plugin/iasyncplugin.md)** (remember to refrence [Flow.Launcher.Plugin](https://www.nuget.org/packages/Flow.Launcher.Plugin/) by Nuget). The plugin template will add the reference and create  a`Main.cs` that implement `IPlugin`.
+
+Found our API Reference [here](API-Reference/)
+
+
+A sample CSharp Plugin [here](https://github.com/Flow-Launcher/plugin-samples/tree/master/HelloWorldCSharp)
```
Comment (JohnTheGr8) [inline develop-csharp-plugins.md:16]: This link could point to the plugin repository, and could even be moved to the sidebar

File: develop-csharp-plugins.md [hunk 12]
```diff
@@ -1,7 +1,71 @@
...
+2. A Dotnet class that implement **[IPlugin](API-Reference/Flow.Launcher.Plugin/iplugin.md)** or **[IAsyncPlugin](API-Reference/Flow.Launcher.Plugin/iasyncplugin.md)** (remember to refrence [Flow.Launcher.Plugin](https://www.nuget.org/packages/Flow.Launcher.Plugin/) by Nuget). The plugin template will add the reference and create  a`Main.cs` that implement `IPlugin`.
+
+Found our API Reference [here](API-Reference/)
+
+
+A sample CSharp Plugin [here](https://github.com/Flow-Launcher/plugin-samples/tree/master/HelloWorldCSharp) 
+
+## IPlugin/IAsyncPlugin
+
+The `Main`class that implement **[IPlugin](API-Reference/Flow.Launcher.Plugin/iplugin.md)** or **[IAsyncPlugin](API-Reference/Flow.Launcher.Plugin/iasyncplugin.md)** will handle the query search with Flow.
```
Comment (JohnTheGr8) [inline develop-csharp-plugins.md:20]: "implement" -> "implements"

File: develop-csharp-plugins.md [hunk 13]
```diff
@@ -1,7 +1,71 @@
...
+
+**[IPlugin](API-Reference/Flow.Launcher.Plugin/iplugin.md)** interface contains two required method
```
Comment (JohnTheGr8) [inline develop-csharp-plugins.md:22]: Add a "The" at the start. Also "method" -> "methods:"

File: develop-csharp-plugins.md [hunk 14]
```diff
@@ -1,7 +1,71 @@
...
+1. `void Init(PluginInitContext context)`
+    - [PluginInitContext](https://github.com/Flow-Launcher/Flow.Launcher/blob/master/API-Reference/Flow.Launcher.Plugin/PluginInitContext.cs) exposes some API from Flow and an metadata object for your plugin. 
+    - `Init`method will be invoked before the invocation of `Query`, so you can do some preparation here.
```
Comment (JohnTheGr8) [inline develop-csharp-plugins.md:25]: "`Init`method will be " can be omitted

File: develop-csharp-plugins.md [hunk 15]
```diff
@@ -1,7 +1,71 @@
...
+    - A `List`of [Result](/API-Reference/Flow.Launcher.Plugin/result.md) object should be returned.
+ 
+ **[IAsyncPlugin](API-Reference/Flow.Launcher.Plugin/iasyncplugin.md)** is the async version of **[IPlugin](API-Reference/Flow.Launcher.Plugin/iplugin.md)**
+ - Instead of implmenting `Init` and `Query`, you will need to implement `InitAsync`and `QueryAsync`, which use `Task`,`Task<List<Result>` as return value to allow  using `async/await`strategy
+ - `QueryAsync` provides a `CancellationToken token` to allow you to check whether user has typed a new query.
+
+
+## Additional Interface
+
+Besides the basic implementation of **IPlugin/IAsyncPlugin**, plugins can also implement a series of interface that belongs to **IFeatures** to control more communication with Flow.
```
Comment (JohnTheGr8) [inline develop-csharp-plugins.md:38]: "a series of interface that belongs to" -> "a series of interfaces that belong to"

File: develop-csharp-plugins.md [hunk 16]
```diff
@@ -1,7 +1,71 @@
...
+ **[IAsyncPlugin](API-Reference/Flow.Launcher.Plugin/iasyncplugin.md)** is the async version of **[IPlugin](API-Reference/Flow.Launcher.Plugin/iplugin.md)**
+ - Instead of implmenting `Init` and `Query`, you will need to implement `InitAsync`and `QueryAsync`, which use `Task`,`Task<List<Result>` as return value to allow  using `async/await`strategy
+ - `QueryAsync` provides a `CancellationToken token` to allow you to check whether user has typed a new query.
+
+
+## Additional Interface
+
+Besides the basic implementation of **IPlugin/IAsyncPlugin**, plugins can also implement a series of interface that belongs to **IFeatures** to control more communication with Flow. 
+
+**Remarks**: We requires you to implement these interfaces in the same class that implement **IPlugin/IAsyncPlugin**.
```
Comment (JohnTheGr8) [inline develop-csharp-plugins.md:40]: "We requires you to" -> "You should"
"same class that implement" -> "same class that implements"

File: develop-csharp-plugins.md [hunk 17]
```diff
@@ -1,7 +1,71 @@
...
+`LoadContextMenus` will be invoked when users expand the context menu of a specific Result. 
+The return value of `LoadContextMenus` is similar to Results from`Query/QueryAsync`.
+
+### [IReloadable](API-Reference/Flow.Launcher.Plugin/ireloadable.md)/[IAsyncReloadable](API-Reference/Flow.Launcher.Plugin/iasyncreloadable.md)
+
+`ReloadData/ReloadDataAsync` will be invoked when users click the `Reload Plugin Data` command from _sys_ plugin. Generally, it is used to reload some cache (such as the programs information cached in _Program_ plugin).
+
+### [IPluginI18n](/API-Reference/Flow.Launcher.Plugin/iplugini18n.md)
+
+**IPluginI18n** means the plugins has been internationalized. Therefore, Flow will load the additional lauguage resources from `/Language` when loading the plugin.
```
Comment (JohnTheGr8) [inline develop-csharp-plugins.md:53]: "plugins" -> "plugin"

Also isn't it `/Languages` instead of `/Language`?

File: develop-csharp-plugins.md [hunk 18]
```diff
@@ -1,7 +1,71 @@
...
+By implementing this interface with additional language files, Flow will be able to load localized language resources. You will able to get the translated text with `IPublicAPI.GetTranslation(string key)`.
```
Comment (JohnTheGr8) [inline develop-csharp-plugins.md:54]: "You will able" -> "You will be able"

File: develop-csharp-plugins.md [hunk 19]
```diff
@@ -1,7 +1,71 @@
...
+
+#### Language Resource
+
+A Language Resource file will have name of the specific Language Code with suffix `.xaml`. The information of the Language Code can be found here [AvailableLanguages.cs](https://github.com/Flow-Launcher/Flow.Launcher/blob/dev/Flow.Launcher.Core/Resource/AvailableLanguages.cs).
+The Language Resource file will need to be a list of **key/value** set. Follow the examples found [here](https://github.com/Flow-Launcher/Flow.Launcher/blob/dev/Flow.Launcher/Languages/en.xaml).
```
Comment (JohnTheGr8) [inline develop-csharp-plugins.md:59]: "key/value set" -> "key/value pairs"

File: develop-csharp-plugins.md [hunk 20]
```diff
@@ -1,7 +1,71 @@
...
+
+#### Remark
+
+Plugins are required to implement **IPublicI18n** to let Flow loads Language resources.
```
Comment (JohnTheGr8) [inline develop-csharp-plugins.md:63]: "loads" -> "load"

File: develop-csharp-plugins.md [hunk 21]
```diff
@@ -1,7 +1,71 @@
...
+
+### [IResultUpdated](API-Reference/Flow.Launcher.Plugin/iresultupdated.md)
+
+
+Implementing **IResultUpdated** provides a way to early update part of the results when you have long running query. This is generally used when plugins want to return some local results earlier than remote results.
```
Comment (JohnTheGr8) [inline develop-csharp-plugins.md:68]: Phrasing is a bit odd here. Maybe something like:

> Provides a way to return part of the query results early. This is generally useful for plugins with long running queries.

File: develop-csharp-plugins.md [hunk 22]
```diff
@@ -1,7 +1,71 @@
...
+
+To early invoke a result to Flow, you will need to invoke `ResultUpdated` event with an `ResultUpdatedEventArgs`, which includes the current `Query` object and the List of `Result` objects similar to the return value in `Query(Async)`.
```
Comment (JohnTheGr8) [inline develop-csharp-plugins.md:70]: "To early invoke a result to Flow" -> "To early return results to Flow"

**Review by JohnTheGr8 — COMMENTED** [review]
A few more minor (typo/wording/etc) corrections. I also think "Develop C# Plugins" in the sidebar should be above "Develop Python Plugins" since it's the more standard way...

**Comment by Zeroto521** [comment]
> By the way, I think vue theme looks cleaner than bubble theme. You can check the effect here.
> https://taooceros.github.io/docs/#/

I don't mind using another style, but better do it at another pr.

File: develop-csharp-plugins.md (hunk 11 above)
Comment (taooceros) [reply develop-csharp-plugins.md:16]: > This link could point to the plugin repository, and could even be moved to the sidebar

What do you mean by plugin repository?

Comment (JohnTheGr8) [reply develop-csharp-plugins.md]: Sorry, I meant the plugin samples repository

**Review by jjw24 — APPROVED** [review]

**Pull request merged**